File: satellites/swift/NovaCompanionKit/Sources/NovaCompanionKit/Protocol.swift
import Foundation

/// Companion protocol v1, mirroring `src/nova_voice/companion/protocol.py`.
///
/// The two sides are kept in agreement by fixtures rather than by a code
/// generator: the server writes `docs/companion-protocol/*.json` from its own
/// models, and this package's tests decode those exact files. A change on
/// either side that the other cannot parse fails a test instead of failing in
/// a room at two in the morning.
public enum CompanionProtocol {
    public static let version = 1
    /// Frames larger than this are rejected before parsing on the server, so
    /// the client must not build one either.
    public static let maxFrameBytes = 256 * 1024
}

public enum CompanionWorkload: String, Codable, Sendable, CaseIterable {
    case interpret
    case renderResponse = "render_response"
    case confirmObjective = "confirm_objective"
    case extractSelfProfileUpdate = "extract_self_profile_update"
    case classifyIcon = "classify_icon"
}

public enum Sensitivity: String, Codable, Sendable {
    case ordinary, personal, location, health, mutation
}

public enum Locality: String, Codable, Sendable {
    case homeLan = "home_lan"
    case tailnet
    case other
}

public enum CompanionRole: String, Codable, Sendable {
    case companion, satellite
}

public enum RejectReason: String, Codable, Sendable {
    case battery, thermal, memory, busy, disabled
    case modelUnavailable = "model_unavailable"
    case unsupportedWorkload = "unsupported_workload"
    case unsupportedSchema = "unsupported_schema"
    case deadlineTooShort = "deadline_too_short"
}

public enum FailureReason: String, Codable, Sendable {
    case timeout, cancelled, internalError = "internal"
    case modelError = "model_error"
    case invalidResult = "invalid_result"
    case toolLimit = "tool_limit"
}

public enum ThermalState: String, Codable, Sendable {
    case nominal, fair, serious, critical
}

public enum AppState: String, Codable, Sendable {
    case foreground, background, inactive
}

public enum PermissionStatus: String, Codable, Sendable {
    case available, denied, restricted, partial, unknown
}

// MARK: - Envelope

/// Identity and lifetime carried by every job-related frame.
///
/// `attemptId` is what makes a late result detectable: a job may be retried,
/// each offer creates a new attempt, and a frame naming a stale one is ignored.
/// `idempotencyKey` is stable across attempts *and* across a local fallback,
/// so a retry cannot duplicate a downstream write.
public struct JobEnvelope: Codable, Sendable {
    public let jobId: String
    public let attemptId: String
    public let idempotencyKey: String
    public let workload: CompanionWorkload
    public let inputRevision: String
    public let schemaVersion: Int
    public let resultSchema: String
    public let sensitivity: Sensitivity
    public let locality: Locality
    public let traceId: String
    public let createdAt: Date
    public let acceptDeadline: Date
    public let completeDeadline: Date
}

// MARK: - Authentication

public struct AuthChallenge: Codable, Sendable {
    public let nonce: String
    public let supportedVersions: [Int]
    public let expiresAt: Date
}

public struct AuthResponse: Codable, Sendable {
    public let protocolVersion: Int
    public let announcedId: String
    public let roles: [CompanionRole]
    public let certificateChain: [String]
    public let signature: String

    public init(
        protocolVersion: Int = CompanionProtocol.version,
        announcedId: String,
        roles: [CompanionRole],
        certificateChain: [String],
        signature: String
    ) {
        self.protocolVersion = protocolVersion
        self.announcedId = announcedId
        self.roles = roles
        self.certificateChain = certificateChain
        self.signature = signature
    }
}

public struct HelloAck: Codable, Sendable {
    public let protocolVersion: Int
    public let authenticatedId: String
    public let locality: Locality
    public let sessionId: String
    public let heartbeatSeconds: Double
}

// MARK: - Capability and runtime state

public struct ModelAvailability: Codable, Sendable {
    public var hotAvailable: Bool
    public var hotContextTokens: Int
    public var deepAvailable: Bool
    public var deepModelId: String?

    public init(
        hotAvailable: Bool = false,
        hotContextTokens: Int = 4096,
        deepAvailable: Bool = false,
        deepModelId: String? = nil
    ) {
        self.hotAvailable = hotAvailable
        self.hotContextTokens = hotContextTokens
        self.deepAvailable = deepAvailable
        self.deepModelId = deepModelId
    }
}

public struct PermissionState: Codable, Sendable {
    public var calendar: PermissionStatus
    public var reminders: PermissionStatus
    public var location: PermissionStatus
    public var health: PermissionStatus

    public init(
        calendar: PermissionStatus = .unknown,
        reminders: PermissionStatus = .unknown,
        location: PermissionStatus = .unknown,
        health: PermissionStatus = .unknown
    ) {
        self.calendar = calendar
        self.reminders = reminders
        self.location = location
        self.health = health
    }
}

public struct CompanionTelemetry: Codable, Sendable {
    public var battery: Double
    public var charging: Bool
    public var lowPowerMode: Bool
    public var thermalState: ThermalState
    public var appState: AppState
    public var models: ModelAvailability
    public var permissions: PermissionState
    /// Diagnostic only. The server classifies locality from the peer address;
    /// a device claiming to be home cannot unlock a home-LAN route by saying so.
    public var reportedNetwork: String?

    public init(
        battery: Double = 1,
        charging: Bool = false,
        lowPowerMode: Bool = false,
        thermalState: ThermalState = .nominal,
        appState: AppState = .background,
        models: ModelAvailability = ModelAvailability(),
        permissions: PermissionState = PermissionState(),
        reportedNetwork: String? = nil
    ) {
        self.battery = battery
        self.charging = charging
        self.lowPowerMode = lowPowerMode
        self.thermalState = thermalState
        self.appState = appState
        self.models = models
        self.permissions = permissions
        self.reportedNetwork = reportedNetwork
    }
}

public struct CompanionHello: Codable, Sendable {
    public let protocolVersion: Int
    public let schemaVersions: [Int]
    public let displayName: String
    public let roles: [CompanionRole]
    public let appVersion: String
    public let osVersion: String
    public let workloads: [CompanionWorkload]
    public let personalTools: [String]
    public let telemetry: CompanionTelemetry
    /// Durable jobs this device believes it is still working on.
    ///
    /// Empty after a restart — which is the point. The server compares this
    /// with what it thinks it leased, so an empty claim from a device that
    /// crashed mid-job is how that job gets released rather than waiting out
    /// its lease.
    ///
    /// **Optional, and omitted when there is nothing to claim.** The server's
    /// message models are `extra="forbid"`, so a hello carrying a field an
    /// older server does not know is rejected and the socket closes. Sending
    /// it unconditionally broke a live device against a server whose matching
    /// change had not shipped yet — the client updates the moment it is
    /// installed, the server updates on a deploy, and the two are not the same
    /// moment. Omitting the empty case means the field only appears once there
    /// is something to say, by which time the server understands it.
    public let activeJobs: [String]?
}

// MARK: - Job lifecycle

public struct JobOffer: Codable, Sendable {
    public let envelope: JobEnvelope
    public let payload: JSONValue
    public let callbackBudget: Int
    public let contextTokens: Int
    /// Exactly the tools this attempt may call back for, as `provider.tool`.
    ///
    /// The device is told the catalogue rather than left to infer it from the
    /// prompt. Nothing here is a security boundary — Iridium refuses anything
    /// outside the list regardless of what the device believes — but naming a
    /// tool that will certainly be refused wastes a round trip and a slice of
    /// the callback budget, so the runtime should read this rather than guess.
    ///
    /// Optional because these arrived after the first committed fixtures: an
    /// offer from an older server decodes with them absent rather than
    /// failing the whole frame.
    public let toolCatalogue: [String]?
    public let callbackDeadlineSeconds: Double?
    public let callbackBudgetSeconds: Double?
    public let maxConcurrentCallbacks: Int?
}

public struct JobAccept: Codable, Sendable {
    public enum Tier: String, Codable, Sendable { case hot, deep }
    public let jobId: String
    public let attemptId: String
    public let tier: Tier
}

public struct JobReject: Codable, Sendable {
    public let jobId: String
    public let attemptId: String
    public let reason: RejectReason
    public let detail: String?
    public let retryAfterSeconds: Double?
}

public struct JobProgress: Codable, Sendable {
    public let jobId: String
    public let attemptId: String
    /// Monotonic within an attempt; a lower sequence is a duplicate.
    public let sequence: Int
    public let stage: String
    public let fraction: Double?
    public let summary: String?
}

public struct JobResult: Codable, Sendable {
    public let jobId: String
    public let attemptId: String
    public let result: JSONValue
    public let tokensUsed: Int?
}

public struct JobFailed: Codable, Sendable {
    public let jobId: String
    public let attemptId: String
    public let reason: FailureReason
    public let detail: String?
}

public struct JobCancel: Codable, Sendable {
    public enum Reason: String, Codable, Sendable {
        case superseded, deadline, shutdown
        case userCancelled = "user_cancelled"
        case localityLost = "locality_lost"
    }
    public let jobId: String
    public let attemptId: String
    public let reason: Reason
}

// MARK: - Tool callbacks

public struct ToolCall: Codable, Sendable {
    public let jobId: String
    public let attemptId: String
    public let callId: String
    public let provider: String
    public let tool: String
    public let arguments: JSONValue
}

public struct ToolResult: Codable, Sendable {
    public let jobId: String
    public let attemptId: String
    public let callId: String
    public let ok: Bool
    public let code: String
    public let message: String
    public let observed: JSONValue?
    public let sensitivity: Sensitivity
}

// MARK: - Approvals

public struct ApprovalRequest: Codable, Sendable {
    public let approvalId: String
    public let jobId: String?
    public let summary: String
    public let provider: String
    public let tool: String
    public let sensitivity: Sensitivity
    public let expiresAt: Date
}

public struct ApprovalResponse: Codable, Sendable {
    public let approvalId: String
    public let approved: Bool
    public let signature: String
}

// MARK: - Personal context

public struct PersonalCall: Codable, Sendable {
    public let callId: String
    public let tool: String
    public let arguments: JSONValue
    public let maxItems: Int
    public let deadlineSeconds: Double
}

public struct PersonalResult: Codable, Sendable {
    public let callId: String
    public let ok: Bool
    public let code: String
    public let message: String
    public let sensitivity: Sensitivity
    public let items: [JSONValue]
    public let truncated: Bool

    /// Explicit and public because the app builds these, and a synthesised
    /// memberwise initialiser on a public struct is only internal.
    public init(
        callId: String,
        ok: Bool,
        code: String,
        message: String,
        sensitivity: Sensitivity,
        items: [JSONValue],
        truncated: Bool
    ) {
        self.callId = callId
        self.ok = ok
        self.code = code
        self.message = message
        self.sensitivity = sensitivity
        self.items = items
        self.truncated = truncated
    }
}

// MARK: - Housekeeping

public struct Heartbeat: Codable, Sendable {
    public let sentAt: Date
    public init(sentAt: Date = Date()) { self.sentAt = sentAt }
}

public struct Ping: Codable, Sendable {
    public let sentAt: Date
}

public struct ConfigurationChanged: Codable, Sendable {
    /// Named settings only — never the values, which may be secret.
    public let changed: [String]
}

public struct TelemetryMessage: Codable, Sendable {
    public let telemetry: CompanionTelemetry
}

public struct LogEvent: Codable, Sendable {
    public enum Level: String, Codable, Sendable { case debug, info, warning, error }
    public let level: Level
    public let event: String
    /// Structural only. Never event titles, reminder notes, locations or
    /// Health values.
    public let detail: String?
    public let traceId: String?
}

File: src/nova_voice/companion/protocol.py
"""Strict, versioned wire contract for the companion channel (protocol v1).

Everything crossing this socket is parsed into one of these models before it is
looked at. ``extra="forbid"`` and explicit bounds are the point: an unknown
message type, an unknown field, an out-of-range deadline or an oversized
payload is rejected deterministically at the edge and never reaches a model, a
capability provider or a durable job row.

Two invariants shape the message set:

* **Offer, then accept.** Iridium *offers* a job; the companion accepts or
  rejects it. Rejection is ordinary flow, not a failure. Once an attempt is
  accepted, Iridium does not run the same workload locally — the whole purpose
  is to free the single llama.cpp slot, and an eager local hedge would defeat
  it. The local path starts only on a terminal outcome for that attempt.
* **The companion plans; Iridium executes.** A ``tool_call`` is a *request*.
  It re-enters Iridium's capability registry, tool policy and household
  authority unchanged, and mutations become durable approval steps. The
  companion holds no dashboard credentials and cannot name a tool that was not
  advertised to it.

Frames are transport events. The durable job row is the authority on state;
progress is structured and monotonic, never parsed back out of log text.
"""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter

PROTOCOL_VERSION = 1
# Protocol versions this server will negotiate. A client outside the range is
# closed at the handshake rather than being half-supported.
SUPPORTED_PROTOCOL_VERSIONS = (1,)

# Hard ceilings enforced before parsing, so a payload bomb costs one length
# check rather than a full JSON parse and model validation.
MAX_FRAME_BYTES = 256 * 1024
MAX_RESULT_BYTES = 128 * 1024
MAX_TOOL_ARGUMENT_BYTES = 32 * 1024

# Workloads that may be routed. Closed so a configured route cannot name a
# workload that has no handler and no schema (NPT-003).
#
# These are exactly the five passes that occupy llama.cpp's single slot today.
# Research synthesis, briefing composition and automation drafting are absent
# on purpose: they are deterministic in this codebase, so there is no local
# implementation for a route to fall back to. Adding them means building that
# local stage first. See nova_voice.companion.workloads and
# docs/COMPANION-BASELINE.md.
CompanionWorkload = Literal[
    "interpret",
    "render_response",
    "confirm_objective",
    "extract_self_profile_update",
    "classify_icon",
]

# How much care a payload needs in logs, storage and retention (NPT-005).
Sensitivity = Literal["ordinary", "personal", "location", "health", "mutation"]

# Where the peer actually connected from, as classified by the server. The
# phone's own opinion is telemetry; this is the security-relevant value.
Locality = Literal["home_lan", "tailnet", "other"]

CompanionRole = Literal["companion", "satellite"]

RejectReason = Literal[
    "battery",
    "thermal",
    "memory",
    "model_unavailable",
    "busy",
    "unsupported_workload",
    "unsupported_schema",
    "deadline_too_short",
    "disabled",
]

FailureReason = Literal[
    "model_error",
    "invalid_result",
    "tool_limit",
    "timeout",
    "cancelled",
    "internal",
]


class CompanionModel(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


# -- envelope -----------------------------------------------------------------


class JobEnvelope(CompanionModel):
    """Identity and lifetime carried by every job-related frame.

    ``attempt_id`` is what makes a late result from a superseded attempt
    detectable: the job may be retried, but each offer creates a new attempt,
    and a frame naming a stale one is recorded as late and ignored.
    ``idempotency_key`` is stable across attempts and across a local fallback,
    so downstream writes cannot be duplicated by a retry.
    """

    job_id: str = Field(alias="jobId", min_length=1, max_length=64)
    attempt_id: str = Field(alias="attemptId", min_length=1, max_length=64)
    idempotency_key: str = Field(alias="idempotencyKey", min_length=1, max_length=128)
    workload: CompanionWorkload
    # Hash/revision of the exact input. A fallback reuses it unchanged.
    input_revision: str = Field(alias="inputRevision", min_length=1, max_length=128)
    schema_version: int = Field(default=1, ge=1, le=1000, alias="schemaVersion")
    result_schema: str = Field(alias="resultSchema", min_length=1, max_length=128)
    sensitivity: Sensitivity = "ordinary"
    locality: Locality = "home_lan"
    # Safe to log: correlates frames, job rows and audit entries without
    # carrying anything about what the request was.
    trace_id: str = Field(alias="traceId", min_length=1, max_length=64)
    created_at: datetime = Field(alias="createdAt")
    accept_deadline: datetime = Field(alias="acceptDeadline")
    complete_deadline: datetime = Field(alias="completeDeadline")


# -- authentication -----------------------------------------------------------


class AuthChallenge(CompanionModel):
    """Server to client: prove you hold the key for the identity you claim."""

    type: Literal["auth_challenge"] = "auth_challenge"
    nonce: str = Field(min_length=32, max_length=128)
    supported_versions: list[int] = Field(alias="supportedVersions")
    expires_at: datetime = Field(alias="expiresAt")


class AuthResponse(CompanionModel):
    type: Literal["auth_response"] = "auth_response"
    protocol_version: int = Field(alias="protocolVersion", ge=1, le=1000)
    announced_id: str = Field(alias="announcedId", min_length=1, max_length=64)
    roles: list[CompanionRole] = Field(min_length=1)
    # PEM leaf first, then any intermediates, validated against the household CA.
    certificate_chain: list[str] = Field(alias="certificateChain", min_length=1, max_length=4)
    # Base64 signature over the canonical challenge string. The signed material
    # covers the nonce, version, id and roles together, so none of them can be
    # swapped after the fact.
    signature: str = Field(min_length=1, max_length=2048)


class HelloAck(CompanionModel):
    type: Literal["hello_ack"] = "hello_ack"
    protocol_version: int = Field(alias="protocolVersion")
    authenticated_id: str = Field(alias="authenticatedId")
    locality: Locality
    session_id: str = Field(alias="sessionId")
    heartbeat_seconds: float = Field(default=20.0, gt=0, le=300, alias="heartbeatSeconds")


# -- capability and runtime state ---------------------------------------------


class ModelAvailability(CompanionModel):
    hot_available: bool = Field(default=False, alias="hotAvailable")
    hot_context_tokens: int = Field(
        default=4096, ge=512, le=1_000_000, alias="hotContextTokens"
    )
    # Optional local weights. Never a prerequisite for reachability.
    deep_available: bool = Field(default=False, alias="deepAvailable")
    deep_model_id: str | None = Field(default=None, max_length=128, alias="deepModelId")


class PermissionState(CompanionModel):
    """Per-capability permission, so one denial degrades only its own tools."""

    calendar: Literal["available", "denied", "restricted", "partial", "unknown"] = "unknown"
    reminders: Literal["available", "denied", "restricted", "partial", "unknown"] = "unknown"
    location: Literal["available", "denied", "restricted", "partial", "unknown"] = "unknown"
    health: Literal["available", "denied", "restricted", "partial", "unknown"] = "unknown"


class CompanionTelemetry(CompanionModel):
    battery: float = Field(default=1.0, ge=0, le=1)
    charging: bool = False
    low_power_mode: bool = Field(default=False, alias="lowPowerMode")
    thermal_state: Literal["nominal", "fair", "serious", "critical"] = Field(
        default="nominal", alias="thermalState"
    )
    app_state: Literal["foreground", "background", "inactive"] = Field(
        default="background", alias="appState"
    )
    models: ModelAvailability = Field(default_factory=ModelAvailability)
    permissions: PermissionState = Field(default_factory=PermissionState)
    # Diagnostic only. The server classifies locality from the actual peer
    # address; a phone claiming to be at home cannot unlock a home-LAN route.
    reported_network: str | None = Field(default=None, max_length=64, alias="reportedNetwork")


class CompanionHello(CompanionModel):
    type: Literal["hello"] = "hello"
    protocol_version: int = Field(alias="protocolVersion")
    schema_versions: list[int] = Field(default_factory=lambda: [1], alias="schemaVersions")
    display_name: str = Field(alias="displayName", min_length=1, max_length=64)
    roles: list[CompanionRole] = Field(min_length=1)
    app_version: str = Field(alias="appVersion", max_length=32)
    os_version: str = Field(alias="osVersion", max_length=32)
    workloads: list[CompanionWorkload] = Field(default_factory=list)
    personal_tools: list[str] = Field(
        default_factory=list, max_length=64, alias="personalTools"
    )
    telemetry: CompanionTelemetry = Field(default_factory=CompanionTelemetry)
    # Durable jobs this device believes it is still working on.
    #
    # Sent on every hello, including the first, where it is empty. Its absence
    # and its emptiness must mean the same thing — "I hold nothing" — because
    # an older client that never sends it has, in fact, restarted and holds
    # nothing. Treating absence as "unknown" instead would leave those jobs
    # leased to a device that will never finish them.
    active_jobs: list[str] = Field(
        default_factory=list, max_length=64, alias="activeJobs"
    )


class TelemetryMessage(CompanionModel):
    type: Literal["telemetry"] = "telemetry"
    telemetry: CompanionTelemetry


class Heartbeat(CompanionModel):
    type: Literal["heartbeat"] = "heartbeat"
    sent_at: datetime = Field(alias="sentAt")


class Ping(CompanionModel):
    type: Literal["ping"] = "ping"
    sent_at: datetime = Field(alias="sentAt")


class ConfigurationChanged(CompanionModel):
    type: Literal["configuration_changed"] = "configuration_changed"
    # Named settings only — never the values, which may be secret.
    changed: list[str] = Field(default_factory=list, max_length=64)


# -- job lifecycle ------------------------------------------------------------


class JobOffer(CompanionModel):
    type: Literal["job_offer"] = "job_offer"
    envelope: JobEnvelope
    payload: dict[str, Any] = Field(default_factory=dict)
    callback_budget: int = Field(default=12, ge=0, le=64, alias="callbackBudget")
    context_tokens: int = Field(default=4096, ge=512, le=1_000_000, alias="contextTokens")
    # Exactly the tools this attempt may call back for, as ``provider.tool``.
    # The companion is told the catalogue rather than left to guess it: a call
    # naming anything outside this list is refused in the session manager,
    # before the registry ever sees it. An empty catalogue means no callbacks.
    tool_catalogue: list[str] = Field(
        default_factory=list, max_length=64, alias="toolCatalogue"
    )
    # Wall-clock ceiling for one callback, and for every callback this attempt
    # makes put together. Count alone is not a bound: twelve calls that each
    # hang for the workload deadline are twelve times the latency the voice
    # turn had budgeted for.
    callback_deadline_seconds: float = Field(
        default=8.0, gt=0, le=120, alias="callbackDeadlineSeconds"
    )
    callback_budget_seconds: float = Field(
        default=30.0, gt=0, le=600, alias="callbackBudgetSeconds"
    )
    max_concurrent_callbacks: int = Field(
        default=2, ge=1, le=8, alias="maxConcurrentCallbacks"
    )


class JobAccept(CompanionModel):
    type: Literal["job_accept"] = "job_accept"
    job_id: str = Field(alias="jobId")
    attempt_id: str = Field(alias="attemptId")
    # Which tier the companion intends to run it on, for later measurement.
    tier: Literal["hot", "deep"] = "hot"


class JobReject(CompanionModel):
    type: Literal["job_reject"] = "job_reject"
    job_id: str = Field(alias="jobId")
    attempt_id: str = Field(alias="attemptId")
    reason: RejectReason
    detail: str | None = Field(default=None, max_length=200)
    # When set, the router holds off re-offering to this session for a while.
    retry_after_seconds: float | None = Field(
        default=None, ge=0, le=3600, alias="retryAfterSeconds"
    )


class JobProgress(CompanionModel):
    type: Literal["job_progress"] = "job_progress"
    job_id: str = Field(alias="jobId")
    attempt_id: str = Field(alias="attemptId")
    # Monotonic within an attempt; a lower sequence than one already seen is a
    # duplicate or reordered frame and is dropped.
    sequence: int = Field(ge=0)
    stage: str = Field(min_length=1, max_length=64)
    fraction: float | None = Field(default=None, ge=0, le=1)
    # Deliberately a summary, not the model's working text.
    summary: str | None = Field(default=None, max_length=400)


class JobResult(CompanionModel):
    type: Literal["job_result"] = "job_result"
    job_id: str = Field(alias="jobId")
    attempt_id: str = Field(alias="attemptId")
    result: dict[str, Any]
    tokens_used: int | None = Field(default=None, ge=0, le=10_000_000, alias="tokensUsed")


class JobFailed(CompanionModel):
    type: Literal["job_failed"] = "job_failed"
    job_id: str = Field(alias="jobId")
    attempt_id: str = Field(alias="attemptId")
    reason: FailureReason
    detail: str | None = Field(default=None, max_length=400)


class JobCancel(CompanionModel):
    type: Literal["job_cancel"] = "job_cancel"
    job_id: str = Field(alias="jobId")
    attempt_id: str = Field(alias="attemptId")
    reason: Literal["superseded", "user_cancelled", "deadline", "shutdown", "locality_lost"]


# -- tool callbacks -----------------------------------------------------------


class ToolCall(CompanionModel):
    type: Literal["tool_call"] = "tool_call"
    job_id: str = Field(alias="jobId")
    attempt_id: str = Field(alias="attemptId")
    call_id: str = Field(alias="callId", min_length=1, max_length=64)
    provider: str = Field(min_length=1, max_length=64)
    tool: str = Field(min_length=1, max_length=128)
    arguments: dict[str, Any] = Field(default_factory=dict)


class ToolResultMessage(CompanionModel):
    type: Literal["tool_result"] = "tool_result"
    job_id: str = Field(alias="jobId")
    attempt_id: str = Field(alias="attemptId")
    call_id: str = Field(alias="callId")
    ok: bool
    code: str = Field(max_length=32)
    message: str = Field(max_length=1000)
    observed: dict[str, Any] | None = None
    sensitivity: Sensitivity = "ordinary"


# -- approvals ----------------------------------------------------------------


class ApprovalRequest(CompanionModel):
    """Server to client: a mutation is proposed and needs an explicit yes.

    An approval is not an execution. The mutation runs once, afterwards, under
    the stored idempotency key — so a duplicate tap, a reconnect or a server
    restart between approval and execution cannot produce a second write.
    """

    type: Literal["approval_request"] = "approval_request"
    approval_id: str = Field(alias="approvalId", min_length=1, max_length=64)
    job_id: str | None = Field(default=None, alias="jobId")
    # Exactly what will happen, in language the owner can check at a glance.
    summary: str = Field(min_length=1, max_length=400)
    provider: str = Field(min_length=1, max_length=64)
    tool: str = Field(min_length=1, max_length=128)
    sensitivity: Sensitivity = "mutation"
    expires_at: datetime = Field(alias="expiresAt")


class ApprovalResponse(CompanionModel):
    type: Literal["approval_response"] = "approval_response"
    approval_id: str = Field(alias="approvalId")
    approved: bool
    # Signature over the approval id and decision, so an approval cannot be
    # replayed or forged by anything that merely reached the socket.
    signature: str = Field(min_length=1, max_length=2048)


# -- personal context ---------------------------------------------------------


class PersonalCall(CompanionModel):
    type: Literal["personal_call"] = "personal_call"
    call_id: str = Field(alias="callId", min_length=1, max_length=64)
    tool: str = Field(min_length=1, max_length=128)
    arguments: dict[str, Any] = Field(default_factory=dict)
    # Bounded by construction: a personal read names its window and its cap.
    max_items: int = Field(default=50, ge=1, le=500, alias="maxItems")
    deadline_seconds: float = Field(default=15.0, gt=0, le=120, alias="deadlineSeconds")


class PersonalResult(CompanionModel):
    type: Literal["personal_result"] = "personal_result"
    call_id: str = Field(alias="callId")
    ok: bool
    code: str = Field(max_length=32)
    message: str = Field(max_length=1000)
    # Scoped records may cross to Iridium in v1. They are marked so logging,
    # retention and redaction can treat them differently from ordinary payloads.
    sensitivity: Sensitivity = "personal"
    items: list[dict[str, Any]] = Field(default_factory=list, max_length=500)
    truncated: bool = False


class LogEvent(CompanionModel):
    type: Literal["log_event"] = "log_event"
    level: Literal["debug", "info", "warning", "error"] = "info"
    event: str = Field(min_length=1, max_length=64)
    # Structural only. Raw event titles, reminder notes, locations and Health
    # values must never be put here.
    detail: str | None = Field(default=None, max_length=1000)
    trace_id: str | None = Field(default=None, alias="traceId", max_length=64)


ClientMessage = Annotated[
    AuthResponse
    | CompanionHello
    | TelemetryMessage
    | Heartbeat
    | JobAccept
    | JobReject
    | JobProgress
    | JobResult
    | JobFailed
    | ToolCall
    | ApprovalResponse
    | PersonalResult
    | LogEvent,
    Field(discriminator="type"),
]

ServerMessage = Annotated[
    AuthChallenge
    | HelloAck
    | JobOffer
    | JobCancel
    | ToolResultMessage
    | ApprovalRequest
    | PersonalCall
    | ConfigurationChanged
    | Ping,
    Field(discriminator="type"),
]


_CLIENT_MESSAGES = TypeAdapter(ClientMessage)


def parse_client_message(raw: str | bytes) -> ClientMessage:
    """Parse one client frame, rejecting oversized input before validation.

    The size check comes first deliberately: a payload bomb should cost one
    length comparison, not a full JSON parse followed by model validation.
    """

    if len(raw) > MAX_FRAME_BYTES:
        raise ValueError(f"companion frame exceeds {MAX_FRAME_BYTES} bytes")
    return _CLIENT_MESSAGES.validate_json(raw)


def serialize(message) -> dict:
    return message.model_dump(mode="json", by_alias=True)
